exporter/opensearchexporter/trace_bulk_indexer.go

```diff
@@ -0,0 +1,245 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package opensearchexporter // import "github.com/open-telemetry/opentelemetry-collector-contrib/exporter/opensearchexporter"
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"strings"
+	"time"
+
+	"github.com/opensearch-project/opensearch-go/v2"
+	"github.com/opensearch-project/opensearch-go/v2/opensearchutil"
+	"go.opentelemetry.io/collector/consumer/consumererror"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+	"go.uber.org/multierr"
+)
+
+type traceBulkIndexer struct {
+	dataset     string
+	namespace   string
+	errs        []error
+	bulkIndexer opensearchutil.BulkIndexer
+}
+
+func newTraceBulkIndexer(dataset string, namespace string) *traceBulkIndexer {
+	return &traceBulkIndexer{dataset, namespace, nil, nil}
+}
+
+func (tbi *traceBulkIndexer) JoinedError() error {
+	return multierr.Combine(tbi.errs...)
+}
+
+func (tbi *traceBulkIndexer) Start(client *opensearch.Client) error {
+	var startErr error
+	tbi.bulkIndexer, startErr = newOpenSearchBulkIndexer(client, tbi.OnIndexerError)
+	return startErr
+}
+
+func (tbi *traceBulkIndexer) Close(ctx context.Context) {
+	closeErr := tbi.bulkIndexer.Close(ctx)
+	if closeErr != nil {
+		tbi.errs = append(tbi.errs, closeErr)
+	}
+}
+
+func (tbi *traceBulkIndexer) HasErrors() bool {
+	return len(tbi.errs) > 0
+}
+
+func (tbi *traceBulkIndexer) OnIndexerError(_ context.Context, indexerErr error) {
+	if indexerErr != nil {
+		tbi.errs = append(tbi.errs, consumererror.NewPermanent(indexerErr))
```

Comment:
```suggestion
		tbi.appendPermanentError(consumererror.NewPermanent(indexerErr))
```

I'm not sure `appendPermanentError()` is a worthwhile helper since it saves few keystrokes and is one line, but if you've got it might as well use it.